specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md

```diff
@@ -0,0 +1,752 @@
+
+
+
+## Host Mapping Models of different Storage devices
+
+
+
+
+
+### DELL EMC VMAX 
+VMAX uses masking view to represnt the Host to LUN access mechanism
+
+![](./Resources/vmax_host_mapping.png)
+
+### DELL EMC VNX 
+
+![](./Resources/vnx_host_mapping.png)
+
+### DELL EMC Unity 
+No host group, but support multi-mapping by modifyLun
+![](./Resources/unity_host_mapping.png)
+
+
+### HP 3PAR
+It has VLUN concept(looks like mapping view), support mapping by host/hostgroup and lun/lungroup
+![](./Resources/3par_host_mapping.png)
+
+### IBM SVC
+![](./Resources/svc_host_mapping.png)
+
+### HDS VSP
+Each initiator is considers as host, mutiple HBA is same host may be represented as mutiple hosts. host name is optional. Host Group ( Initiator group) is supported
+![](./Resources/vsp_host_mapping.png)
+
+## Delfin host mapping objects
+### Masking view
+The masking view identifies the host group (set of hosts), the port group (set of array ports), and the storage group (set of volumes) in a combination which are associated. 
+### Host group
+logical grouping of hosts
+### Port group
+Logical grouping of array ports .
+### Storage group
+Logical grouping of volumes.
+### Host
+Represents a host , consists of mutiple initiators
+### Initiator
+An HBA wwn or iscsi traget.
+### Port
+Array port through the volume is mapped to the initiator
+### Volume
+volume or lun or logical device or vitual volume which is mapped to initiators.
+
+## Host mapping object requirement mapping to some storage models
+| Object        | VMAX     | Unity    | HP 3par  | HDS VSP | IBM SVC | EMC VNX                                                         |
+|---------------|----------|----------|----------|---------|---------|-----------------------------------------------------------------|
+| masking_view  | yes      | No       | no       | no      | no      | no                                                              |
+| host          | yes      | yes      | yes      | yes     | yes     | yes                                                             |
+| host_group    | optional | no       | optional | yes     | no      | no                                                              |
+| initiator     | yes      | yes      | yes      | yes     | yes     | yes                                                             |
+| port          | yes      | no       | no       | no      | no      | no                                                              |
+| port_group    | yes      | no       | no       | no      | no      | no                                                              |
+| storage_group | yes      | optional | no       | no      | no      | optional ( vnx storage group is a collection of hosts and luns) |
+| volume        | yes      | yes      | yes      | yes     | yes     | yes                                                             |
+
+## Proposed Delfin common model 
+
+
+![](./Resources/Delfin_host_mapping.png)
```

Comment:
Few questions on the Model:
If a Vendor doesn't support StorageGroup or PortGroup or HostGroup, then will we create a dummy group for Maksing View?
Can we segregate properly the masking and mapping views and then connect?
Keep the first class objects as the entities directly involved in making or mapping. Like: MaskingView will contain 1) Port 2) HBA 3) LUN 
Now all the groups like HostGroup, StorageGroup and PortGroup will have second level relationship.

One catch here is that, MaskingView can get bloated if we keep LUN ID there. To solve it maybe we can keep only the ArrayPort and HBA in the Masking view. Volume/LUN info table will have PortID associated. So all info can be retrieved though SQL Join.
HostGroup/PortGroup/StorageGroup will be under the HBA/ArrayPort/LUN.

Reply:
I will add some explanation with work flows to give more clarity

Reply:
added more ino with workflow..

Reply:
Do we really need: StorageGroupVolumeMapping, MaskingViewStorageGroupMapping, MaskingViewPortGroupMapping,  HostGroupHostMapping, MaskingViewHostGroupMapping ? Will the DB query be slow due to many joins?

Reply:
we should have a way to represent one to many relation, rhese mapping table represent the relation. Volume can be part of mutiple storage groups, simlary host to host group and port to port gorup